src/sbuf.3
.\"-
.\" Copyright (c) 2000 Poul-Henning Kamp and Dag-Erling Coïdan Smørgrav
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD: releng/11.1/share/man/man9/sbuf.9 284557 2015-06-18 16:29:11Z brueffer $
.\"
.Dd March 14, 2015
.Dt SBUF 9
.Os
.Sh NAME
.Nm sbuf ,
.Nm sbuf_new ,
.Nm sbuf_new_auto ,
.Nm sbuf_new_for_sysctl ,
.Nm sbuf_clear ,
.Nm sbuf_get_flags ,
.Nm sbuf_set_flags ,
.Nm sbuf_clear_flags ,
.Nm sbuf_setpos ,
.Nm sbuf_bcat ,
.Nm sbuf_bcopyin ,
.Nm sbuf_bcpy ,
.Nm sbuf_cat ,
.Nm sbuf_copyin ,
.Nm sbuf_cpy ,
.Nm sbuf_printf ,
.Nm sbuf_vprintf ,
.Nm sbuf_putc ,
.Nm sbuf_set_drain ,
.Nm sbuf_trim ,
.Nm sbuf_error ,
.Nm sbuf_finish ,
.Nm sbuf_data ,
.Nm sbuf_len ,
.Nm sbuf_done ,
.Nm sbuf_delete ,
.Nm sbuf_start_section ,
.Nm sbuf_end_section ,
.Nm sbuf_hexdump
.Nd safe string composition
.Sh SYNOPSIS
.In sys/types.h
.In sbuf.h
.Ft typedef\ int ( sbuf_drain_func ) ( void\ *arg, const\ char\ *data, int\ len ) ;
.Pp
.Ft struct sbuf *
.Fn sbuf_new "struct sbuf *s" "char *buf" "int length" "int flags"
.Ft struct sbuf *
.Fn sbuf_new_auto
.Ft void
.Fn sbuf_clear "struct sbuf *s"
.Ft int
.Fn sbuf_get_flags "struct sbuf *s"
.Ft void
.Fn sbuf_set_flags "struct sbuf *s" "int flags"
.Ft void
.Fn sbuf_clear_flags "struct sbuf *s" "int flags"
.Ft int
.Fn sbuf_setpos "struct sbuf *s" "int pos"
.Ft int
.Fn sbuf_bcat "struct sbuf *s" "const void *buf" "size_t len"
.Ft int
.Fn sbuf_bcopyin "struct sbuf *s" "const void *uaddr" "size_t len"
.Ft int
.Fn sbuf_bcpy "struct sbuf *s" "const void *buf" "size_t len"
.Ft int
.Fn sbuf_cat "struct sbuf *s" "const char *str"
.Ft int
.Fn sbuf_copyin "struct sbuf *s" "const void *uaddr" "size_t len"
.Ft int
.Fn sbuf_cpy "struct sbuf *s" "const char *str"
.Ft int
.Fn sbuf_printf "struct sbuf *s" "const char *fmt" "..."
.Ft int
.Fn sbuf_vprintf "struct sbuf *s" "const char *fmt" "va_list ap"
.Ft int
.Fn sbuf_putc "struct sbuf *s" "int c"
.Ft void
.Fn sbuf_set_drain "struct sbuf *s" "sbuf_drain_func *func" "void *arg"
.Ft int
.Fn sbuf_trim "struct sbuf *s"
.Ft int
.Fn sbuf_error "struct sbuf *s"
.Ft int
.Fn sbuf_finish "struct sbuf *s"
.Ft char *
.Fn sbuf_data "struct sbuf *s"
.Ft ssize_t
.Fn sbuf_len "struct sbuf *s"
.Ft int
.Fn sbuf_done "struct sbuf *s"
.Ft void
.Fn sbuf_delete "struct sbuf *s"
.Ft void
.Fn sbuf_start_section "struct sbuf *s" "ssize_t *old_lenp"
.Ft ssize_t
.Fn sbuf_end_section "struct sbuf *s" "ssize_t old_len" "size_t pad" "int c"
.Ft void
.Fo sbuf_hexdump
.Fa "struct sbuf *sb"
.Fa "void *ptr"
.Fa "int length"
.Fa "const char *hdr"
.Fa "int flags"
.Fc
.In sys/sysctl.h
.Ft struct sbuf *
.Fn sbuf_new_for_sysctl "struct sbuf *s" "char *buf" "int length" "struct sysctl_req *req"
.Sh DESCRIPTION
The
.Nm
family of functions allows one to safely allocate, compose and
release strings in kernel or user space.
.Pp
Instead of arrays of characters, these functions operate on structures
called
.Fa sbufs ,
defined in
.In sys/sbuf.h .
.Pp
Any errors encountered during the allocation or composition of the
string will be latched in the data structure,
making a single error test at the end of the composition
sufficient to determine success or failure of the entire process.
.Pp
The
.Fn sbuf_new
function initializes the
.Fa sbuf
pointed to by its first argument.
If that pointer is
.Dv NULL ,
.Fn sbuf_new
allocates a
.Vt struct sbuf
using
.Xr calloc 3 .
The
.Fa buf
argument is a pointer to a buffer in which to store the actual string;
if it is
.Dv NULL ,
.Fn sbuf_new
will allocate one using
.Xr calloc 3 .
The
.Fa length
is the initial size of the storage buffer.
The fourth argument,
.Fa flags ,
may be comprised of the following flags:
.Bl -tag -width ".Dv SBUF_AUTOEXTEND"
.It Dv SBUF_FIXEDLEN
The storage buffer is fixed at its initial size.
Attempting to extend the sbuf beyond this size results in an overflow condition.
.It Dv SBUF_AUTOEXTEND
This indicates that the storage buffer may be extended as necessary, so long
as resources allow, to hold additional data.
.It Dv SBUF_INCLUDENUL
This causes the final nulterm byte to be counted in the length of the data.
.El
.Pp
Note that if
.Fa buf
is not
.Dv NULL ,
it must point to an array of at least
.Fa length
characters.
The result of accessing that array directly while it is in use by the
sbuf is undefined.
.Pp
The
.Fn sbuf_new_auto
function is a shortcut for creating a completely dynamic
.Nm .
It is the equivalent of calling
.Fn sbuf_new
with values
.Dv NULL ,
.Dv NULL ,
.Dv 0 ,
and
.Dv SBUF_AUTOEXTEND .
.Pp
The
.Fn sbuf_delete
function clears the
.Fa sbuf
and frees any memory allocated for it.
There must be a call to
.Fn sbuf_delete
for every call to
.Fn sbuf_new .
Any attempt to access the sbuf after it has been deleted will fail.
.Pp
The
.Fn sbuf_clear
function invalidates the contents of the
.Fa sbuf
and resets its position to zero.
.Pp
The
.Fn sbuf_get_flags
function returns the current user flags.
The
.Fn sbuf_set_flags
and
.Fn sbuf_get_flags
functions set or clear one or more user flags, respectively.
The user flags are described under the
.Fn sbuf_new
function.
.Pp
The
.Fn sbuf_setpos
function sets the
.Fa sbuf Ns 's
end position to
.Fa pos ,
which is a value between zero and one less than the size of the
storage buffer.
This effectively truncates the sbuf at the new position.
.Pp
The
.Fn sbuf_bcat
function appends the first
.Fa len
bytes from the buffer
.Fa buf
to the
.Fa sbuf .
.Pp
The
.Fn sbuf_bcopyin
function copies
.Fa len
bytes from the specified userland address into the
.Fa sbuf .
.Pp
The
.Fn sbuf_bcpy
function replaces the contents of the
.Fa sbuf
with the first
.Fa len
bytes from the buffer
.Fa buf .
.Pp
The
.Fn sbuf_cat
function appends the NUL-terminated string
.Fa str
to the
.Fa sbuf
at the current position.
.Pp
The
.Fn sbuf_set_drain
function sets a drain function
.Fa func
for the
.Fa sbuf ,
and records a pointer
.Fa arg
to be passed to the drain on callback.
The drain function cannot be changed while
.Fa sbuf_len
is non-zero.
.Pp
The registered drain function
.Vt sbuf_drain_func
will be called with the argument
.Fa arg
provided to
.Fn sbuf_set_drain ,
a pointer
.Fa data
to a byte string that is the contents of the sbuf, and the length
.Fa len
of the data.
If the drain function exists, it will be called when the sbuf internal
buffer is full, or on behalf of
.Fn sbuf_finish .
The drain function may drain some or all of the data, but must drain
at least 1 byte.
The return value from the drain function, if positive, indicates how
many bytes were drained.
If negative, the return value indicates the negative error code which
will be returned from this or a later call to
.Fn sbuf_finish .
The returned drained length cannot be zero.
To do unbuffered draining, initialize the sbuf with a two-byte buffer.
The drain will be called for every byte added to the sbuf.
The
.Fn sbuf_bcopyin ,
.Fn sbuf_copyin ,
.Fn sbuf_trim ,
and
.Fn sbuf_data
functions cannot be used on an sbuf with a drain.
.Pp
The
.Fn sbuf_copyin
function copies a NUL-terminated string from the specified userland
address into the
.Fa sbuf .
If the
.Fa len
argument is non-zero, no more than
.Fa len
characters (not counting the terminating NUL) are copied; otherwise
the entire string, or as much of it as can fit in the
.Fa sbuf ,
is copied.
.Pp
The
.Fn sbuf_cpy
function replaces the contents of the
.Fa sbuf
with those of the NUL-terminated string
.Fa str .
This is equivalent to calling
.Fn sbuf_cat
with a fresh
.Fa sbuf
or one which position has been reset to zero with
.Fn sbuf_clear
or
.Fn sbuf_setpos .
.Pp
The
.Fn sbuf_printf
function formats its arguments according to the format string pointed
to by
.Fa fmt
and appends the resulting string to the
.Fa sbuf
at the current position.
.Pp
The
.Fn sbuf_vprintf
function behaves the same as
.Fn sbuf_printf
except that the arguments are obtained from the variable-length argument list
.Fa ap .
.Pp
The
.Fn sbuf_putc
function appends the character
.Fa c
to the
.Fa sbuf
at the current position.
.Pp
The
.Fn sbuf_trim
function removes trailing whitespace from the
.Fa sbuf .
.Pp
The
.Fn sbuf_error
function returns any error value that the
.Fa sbuf
may have accumulated, either from the drain function, or ENOMEM if the
.Fa sbuf
overflowed.
This function is generally not needed and instead the error code from
.Fn sbuf_finish
is the preferred way to discover whether an sbuf had an error.
.Pp
The
.Fn sbuf_finish
function will call the attached drain function if one exists until all
the data in the
.Fa sbuf
is flushed.
If there is no attached drain,
.Fn sbuf_finish
NUL-terminates the
.Fa sbuf .
In either case it marks the
.Fa sbuf
as finished, which means that it may no longer be modified using
.Fn sbuf_setpos ,
.Fn sbuf_cat ,
.Fn sbuf_cpy ,
.Fn sbuf_printf
or
.Fn sbuf_putc ,
until
.Fn sbuf_clear
is used to reset the sbuf.
.Pp
The
.Fn sbuf_data
function returns the actual string;
.Fn sbuf_data
only works on a finished
.Fa sbuf .
The
.Fn sbuf_len
function returns the length of the string.
For an
.Fa sbuf
with an attached drain,
.Fn sbuf_len
returns the length of the un-drained data.
.Fn sbuf_done
returns non-zero if the
.Fa sbuf
is finished.
.Pp
The
.Fn sbuf_start_section
and
.Fn sbuf_end_section
functions may be used for automatic section alignment.
The arguments
.Fa pad
and
.Fa c
specify the padding size and a character used for padding.
The arguments
.Fa old_lenp
and
.Fa old_len
are to save and restore the current section length when nested sections
are used.
For the top level section
.Dv NULL
and \-1 can be specified for
.Fa old_lenp
and
.Fa old_len
respectively.
.Pp
The
.Fn sbuf_hexdump
function prints an array of bytes to the supplied sbuf, along with an ASCII
representation of the bytes if possible.
See the
.Xr hexdump 3
man page for more details on the interface.
.Sh NOTES
If an operation caused an
.Fa sbuf
to overflow, most subsequent operations on it will fail until the
.Fa sbuf
is finished using
.Fn sbuf_finish
or reset using
.Fn sbuf_clear ,
or its position is reset to a value between 0 and one less than the
size of its storage buffer using
.Fn sbuf_setpos ,
or it is reinitialized to a sufficiently short string using
.Fn sbuf_cpy .
.Pp
Drains in user-space will not always function as indicated.
While the drain function will be called immediately on overflow from
the
.Fa sbuf_putc ,
.Fa sbuf_bcat ,
.Fa sbuf_cat
functions,
.Fa sbuf_printf
and
.Fa sbuf_vprintf
currently have no way to determine whether there will be an overflow
until after it occurs, and cannot do a partial expansion of the format
string.
Thus when using libsbuf the buffer may be extended to allow completion
of a single printf call, even though a drain is attached.
.Sh RETURN VALUES
The
.Fn sbuf_new
function returns
.Dv NULL
if it failed to allocate a storage buffer, and a pointer to the new
.Fa sbuf
otherwise.
.Pp
The
.Fn sbuf_setpos
function returns \-1 if
.Fa pos
was invalid, and zero otherwise.
.Pp
The
.Fn sbuf_cat ,
.Fn sbuf_cpy ,
.Fn sbuf_printf ,
.Fn sbuf_putc ,
and
.Fn sbuf_trim
functions
all return \-1 if the buffer overflowed, and zero otherwise.
.Pp
The
.Fn sbuf_error
function returns a non-zero value if the buffer has an overflow or
drain error, and zero otherwise.
.Pp
The
.Fn sbuf_len

src/sbuf.c
/*-
 * Copyright (c) 2000-2008 Poul-Henning Kamp
 * Copyright (c) 2000-2008 Dag-Erling Coïdan Smørgrav
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer
 *    in this position and unchanged.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>

#include <sys/param.h>

#include <ctype.h>
#include <errno.h>
#include <limits.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "sbuf.h"

#define	KASSERT(e, m)
#define	SBMALLOC(size)		calloc(1, size)
#define	SBFREE(buf)		free(buf)

/* 
 * See sys/param.h from FreeBSD 11{.x}
 * 
 *  if y is powers of two 
 */
#define	roundup2(x, y)	(((x)+((y)-1))&(~((y)-1))) 

/*
 * Predicates
 */
#define	SBUF_ISDYNAMIC(s)	((s)->s_flags & SBUF_DYNAMIC)
#define	SBUF_ISDYNSTRUCT(s)	((s)->s_flags & SBUF_DYNSTRUCT)
#define	SBUF_ISFINISHED(s)	((s)->s_flags & SBUF_FINISHED)
#define	SBUF_HASROOM(s)		((s)->s_len < (s)->s_size - 1)
#define	SBUF_FREESPACE(s)	((s)->s_size - ((s)->s_len + 1))
#define	SBUF_CANEXTEND(s)	((s)->s_flags & SBUF_AUTOEXTEND)
#define	SBUF_ISSECTION(s)	((s)->s_flags & SBUF_INSECTION)
#define	SBUF_NULINCLUDED(s)	((s)->s_flags & SBUF_INCLUDENUL)

/*
 * Set / clear flags
 */
#define	SBUF_SETFLAG(s, f)	do { (s)->s_flags |= (f); } while (0)
#define	SBUF_CLEARFLAG(s, f)	do { (s)->s_flags &= ~(f); } while (0)

#define	SBUF_MINSIZE		 2		/* Min is 1 byte + nulterm. */
#define	SBUF_MINEXTENDSIZE	16		/* Should be power of 2. */

static int 	sbuf_extendsize(int);
static int 	sbuf_extend(struct sbuf *, int);
static struct sbuf * 	sbuf_newbuf(struct sbuf *, char *, int, int);
static int 	sbuf_drain(struct sbuf *);
static void 	sbuf_put_bytes(struct sbuf *, const char *, size_t);
static void 	sbuf_put_byte(struct sbuf *, char);

/*
 * Debugging support
 */
#define	assert_sbuf_integrity(s) do { } while (0)
#define	assert_sbuf_state(s, i)	 do { } while (0)

static int
sbuf_extendsize(int size)
{
	int newsize;

	if (size < (int)SBUF_MAXEXTENDSIZE) {
		newsize = SBUF_MINEXTENDSIZE;
		while (newsize < size)
			newsize *= 2;
	} else {
		newsize = roundup2(size, SBUF_MAXEXTENDINCR);
	}
	KASSERT(newsize >= size, ("%s: %d < %d\n", __func__, newsize, size));
	return (newsize);
}

/*
 * Extend an sbuf.
 */
static int
sbuf_extend(struct sbuf *s, int addlen)
{
	char *newbuf;
	int newsize;

	if (!SBUF_CANEXTEND(s))
		return (-1);
	newsize = sbuf_extendsize(s->s_size + addlen);
	newbuf = SBMALLOC(newsize);
	if (newbuf == NULL)
		return (-1);
	memcpy(newbuf, s->s_buf, s->s_size);
	if (SBUF_ISDYNAMIC(s))
		SBFREE(s->s_buf);
	else
		SBUF_SETFLAG(s, SBUF_DYNAMIC);
	s->s_buf = newbuf;
	s->s_size = newsize;
	return (0);
}

/*
 * Initialize the internals of an sbuf.
 * If buf is non-NULL, it points to a static or already-allocated string
 * big enough to hold at least length characters.
 */
static struct sbuf *
sbuf_newbuf(struct sbuf *s, char *buf, int length, int flags)
{

	memset(s, 0, sizeof(*s));
	s->s_flags = flags;
	s->s_size = length;
	s->s_buf = buf;

	if ((s->s_flags & SBUF_AUTOEXTEND) == 0) {
		KASSERT(s->s_size >= SBUF_MINSIZE,
		    ("attempt to create an sbuf smaller than %d bytes",
		    SBUF_MINSIZE));
	}

	if (s->s_buf != NULL)
		return (s);

	if ((flags & SBUF_AUTOEXTEND) != 0)
		s->s_size = sbuf_extendsize(s->s_size);

	s->s_buf = SBMALLOC(s->s_size);
	if (s->s_buf == NULL)
		return (NULL);
	SBUF_SETFLAG(s, SBUF_DYNAMIC);
	return (s);
}

/*
 * Initialize an sbuf.
 * If buf is non-NULL, it points to a static or already-allocated string
 * big enough to hold at least length characters.
 */
struct sbuf *
sbuf_new(struct sbuf *s, char *buf, int length, int flags)
{

	KASSERT(length >= 0,
	    ("attempt to create an sbuf of negative length (%d)", length));
	KASSERT((flags & ~SBUF_USRFLAGMSK) == 0,
	    ("%s called with invalid flags", __func__));

	flags &= SBUF_USRFLAGMSK;
	if (s != NULL)
		return (sbuf_newbuf(s, buf, length, flags));

	s = SBMALLOC(sizeof(*s));
	if (s == NULL)
		return (NULL);
	if (sbuf_newbuf(s, buf, length, flags) == NULL) {
		SBFREE(s);
		return (NULL);
	}
	SBUF_SETFLAG(s, SBUF_DYNSTRUCT);
	return (s);
}

int
sbuf_get_flags(struct sbuf *s)
{

	return (s->s_flags & SBUF_USRFLAGMSK);
}

void
sbuf_clear_flags(struct sbuf *s, int flags)
{

	s->s_flags &= ~(flags & SBUF_USRFLAGMSK);
}

void
sbuf_set_flags(struct sbuf *s, int flags)
{


	s->s_flags |= (flags & SBUF_USRFLAGMSK);
}

/*
 * Clear an sbuf and reset its position.
 */
void
sbuf_clear(struct sbuf *s)
{

	assert_sbuf_integrity(s);
	/* don't care if it's finished or not */

	SBUF_CLEARFLAG(s, SBUF_FINISHED);
	s->s_error = 0;
	s->s_len = 0;
	s->s_sect_len = 0;
}

/*
 * Set the sbuf's end position to an arbitrary value.
 * Effectively truncates the sbuf at the new position.
 */
int
sbuf_setpos(struct sbuf *s, ssize_t pos)
{

	assert_sbuf_integrity(s);
	assert_sbuf_state(s, 0);

	KASSERT(pos >= 0,
	    ("attempt to seek to a negative position (%jd)", (intmax_t)pos));
	KASSERT(pos < s->s_size,
	    ("attempt to seek past end of sbuf (%jd >= %jd)",
	    (intmax_t)pos, (intmax_t)s->s_size));
	KASSERT(!SBUF_ISSECTION(s),
	    ("attempt to seek when in a section"));

	if (pos < 0 || pos > s->s_len)
		return (-1);
	s->s_len = pos;
	return (0);
}

/*
 * Set up a drain function and argument on an sbuf to flush data to
 * when the sbuf buffer overflows.
 */
void
sbuf_set_drain(struct sbuf *s, sbuf_drain_func *func, void *ctx)
{

	assert_sbuf_state(s, 0);
	assert_sbuf_integrity(s);
	KASSERT(func == s->s_drain_func || s->s_len == 0,
	    ("Cannot change drain to %p on non-empty sbuf %p", func, s));
	s->s_drain_func = func;
	s->s_drain_arg = ctx;
}

/*
 * Call the drain and process the return.
 */
static int
sbuf_drain(struct sbuf *s)
{
	int len;

	KASSERT(s->s_len > 0, ("Shouldn't drain empty sbuf %p", s));
	KASSERT(s->s_error == 0, ("Called %s with error on %p", __func__, s));
	len = s->s_drain_func(s->s_drain_arg, s->s_buf, s->s_len);
	if (len < 0) {
		s->s_error = -len;
		return (s->s_error);
	}
	KASSERT(len > 0 && len <= s->s_len,
	    ("Bad drain amount %d for sbuf %p", len, s));
	s->s_len -= len;
	/*
	 * Fast path for the expected case where all the data was
	 * drained.
	 */
	if (s->s_len == 0)
		return (0);
	/*
	 * Move the remaining characters to the beginning of the
	 * string.
	 */
	memmove(s->s_buf, s->s_buf + len, s->s_len);
	return (0);
}

/*
 * Append bytes to an sbuf.  This is the core function for appending
 * to an sbuf and is the main place that deals with extending the
 * buffer and marking overflow.
 */
static void
sbuf_put_bytes(struct sbuf *s, const char *buf, size_t len)
{
	size_t n;

	assert_sbuf_integrity(s);
	assert_sbuf_state(s, 0);

	if (s->s_error != 0)
		return;
	while (len > 0) {
		if (SBUF_FREESPACE(s) <= 0) {
			/*
			 * If there is a drain, use it, otherwise extend the
			 * buffer.
			 */
			if (s->s_drain_func != NULL)
				(void)sbuf_drain(s);
			else if (sbuf_extend(s, len > INT_MAX ? INT_MAX : len)
			    < 0)
				s->s_error = ENOMEM;
			if (s->s_error != 0)
				return;
		}
		n = SBUF_FREESPACE(s);
		if (len < n)
			n = len;
		memcpy(&s->s_buf[s->s_len], buf, n);
		s->s_len += n;
		if (SBUF_ISSECTION(s))
			s->s_sect_len += n;
		len -= n;
		buf += n;
	}
}

static void
sbuf_put_byte(struct sbuf *s, char c)
{

	sbuf_put_bytes(s, &c, 1);
}

/*
 * Append a byte string to an sbuf.
 */
int
sbuf_bcat(struct sbuf *s, const void *buf, size_t len)
{

	sbuf_put_bytes(s, buf, len);
	if (s->s_error != 0)
		return (-1);
	return (0);
}

/*
 * Copy a byte string into an sbuf.
 */
int
sbuf_bcpy(struct sbuf *s, const void *buf, size_t len)
{

	assert_sbuf_integrity(s);
	assert_sbuf_state(s, 0);

	sbuf_clear(s);
	return (sbuf_bcat(s, buf, len));
}

/*
 * Append a string to an sbuf.
 */
int
sbuf_cat(struct sbuf *s, const char *str)
{
	size_t n;

	n = strlen(str);
	sbuf_put_bytes(s, str, n);
	if (s->s_error != 0)
		return (-1);
	return (0);
}


/*
 * Copy a string into an sbuf.
 */
int
sbuf_cpy(struct sbuf *s, const char *str)
{

	assert_sbuf_integrity(s);
	assert_sbuf_state(s, 0);

	sbuf_clear(s);
	return (sbuf_cat(s, str));
}

/*
 * Format the given argument list and append the resulting string to an sbuf.
 */

int
sbuf_vprintf(struct sbuf *s, const char *fmt, va_list ap)
{
	va_list ap_copy;
	int error, len;

	assert_sbuf_integrity(s);
	assert_sbuf_state(s, 0);

	KASSERT(fmt != NULL,
	    ("%s called with a NULL format string", __func__));

	if (s->s_error != 0)
		return (-1);

	/*
	 * For the moment, there is no way to get vsnprintf(3) to hand
	 * back a character at a time, to push everything into
	 * sbuf_putc_func() as was done for the kernel.
	 *
	 * In userspace, while drains are useful, there's generally
	 * not a problem attempting to malloc(3) on out of space.  So
	 * expand a userland sbuf if there is not enough room for the
	 * data produced by sbuf_[v]printf(3).
	 */

	error = 0;
	do {
		va_copy(ap_copy, ap);
		len = vsnprintf(&s->s_buf[s->s_len], SBUF_FREESPACE(s) + 1,
		    fmt, ap_copy);
		if (len < 0) {
			s->s_error = errno;
			return (-1);
		}
		va_end(ap_copy);

		if (SBUF_FREESPACE(s) >= len)
			break;
		/* Cannot print with the current available space. */
		if (s->s_drain_func != NULL && s->s_len > 0)
			error = sbuf_drain(s);
		else
			error = sbuf_extend(s, len - SBUF_FREESPACE(s));
	} while (error == 0);

	/*
	 * s->s_len is the length of the string, without the terminating nul.
	 * When updating s->s_len, we must subtract 1 from the length that
	 * we passed into vsnprintf() because that length includes the
	 * terminating nul.
	 *
	 * vsnprintf() returns the amount that would have been copied,
	 * given sufficient space, so don't over-increment s_len.
	 */
	if (SBUF_FREESPACE(s) < len)
		len = SBUF_FREESPACE(s);
	s->s_len += len;
	if (SBUF_ISSECTION(s))
		s->s_sect_len += len;
	if (!SBUF_HASROOM(s) && !SBUF_CANEXTEND(s))
		s->s_error = ENOMEM;

	KASSERT(s->s_len < s->s_size,
	    ("wrote past end of sbuf (%d >= %d)", s->s_len, s->s_size));

	if (s->s_error != 0)
		return (-1);
	return (0);
}

/*
 * Format the given arguments and append the resulting string to an sbuf.
 */
int
sbuf_printf(struct sbuf *s, const char *fmt, ...)
{
	va_list ap;
	int result;

	va_start(ap, fmt);
	result = sbuf_vprintf(s, fmt, ap);
	va_end(ap);
	return (result);
}

/*
 * Append a character to an sbuf.
 */
int
sbuf_putc(struct sbuf *s, int c)
{

	sbuf_put_byte(s, c);
	if (s->s_error != 0)
		return (-1);
	return (0);
}

/*
 * Trim whitespace characters from end of an sbuf.
 */
int
sbuf_trim(struct sbuf *s)
{

	assert_sbuf_integrity(s);
	assert_sbuf_state(s, 0);
	KASSERT(s->s_drain_func == NULL,
	    ("%s makes no sense on sbuf %p with drain", __func__, s));

	if (s->s_error != 0)
		return (-1);

	while (s->s_len > 0 && isspace(s->s_buf[s->s_len-1])) {
		--s->s_len;
		if (SBUF_ISSECTION(s))
			s->s_sect_len--;
	}

	return (0);
}

/*
 * Check if an sbuf has an error.
 */
int
sbuf_error(const struct sbuf *s)
{

	return (s->s_error);
}

/*
 * Finish off an sbuf.
 */
int
sbuf_finish(struct sbuf *s)
{

	assert_sbuf_integrity(s);
	assert_sbuf_state(s, 0);

	s->s_buf[s->s_len] = '\0';
	if (SBUF_NULINCLUDED(s))
		s->s_len++;
	if (s->s_drain_func != NULL) {
		while (s->s_len > 0 && s->s_error == 0)
			s->s_error = sbuf_drain(s);
	}
	SBUF_SETFLAG(s, SBUF_FINISHED);

	if (s->s_error != 0) {
		errno = s->s_error;
		return (-1);
	}
	return (0);
}

/*
 * Return a pointer to the sbuf data.
 */
char *
sbuf_data(struct sbuf *s)
{

	assert_sbuf_integrity(s);
	assert_sbuf_state(s, SBUF_FINISHED);
	KASSERT(s->s_drain_func == NULL,
	    ("%s makes no sense on sbuf %p with drain", __func__, s));

	return (s->s_buf);
}

/*
 * Return the length of the sbuf data.
 */
ssize_t
sbuf_len(struct sbuf *s)
{

	assert_sbuf_integrity(s);
	/* don't care if it's finished or not */
	KASSERT(s->s_drain_func == NULL,
	    ("%s makes no sense on sbuf %p with drain", __func__, s));

	if (s->s_error != 0)
		return (-1);

	/* If finished, nulterm is already in len, else add one. */
	if (SBUF_NULINCLUDED(s) && !SBUF_ISFINISHED(s))
		return (s->s_len + 1);
	return (s->s_len);
}

/*
 * Clear an sbuf, free its buffer if necessary.
 */
void
sbuf_delete(struct sbuf *s)
{
	int isdyn;

	assert_sbuf_integrity(s);
	/* don't care if it's finished or not */

	if (SBUF_ISDYNAMIC(s))
		SBFREE(s->s_buf);
	isdyn = SBUF_ISDYNSTRUCT(s);
	memset(s, 0, sizeof(*s));
	if (isdyn)
		SBFREE(s);
}

/*
 * Check if an sbuf has been finished.
 */
int
sbuf_done(const struct sbuf *s)
{

	return (SBUF_ISFINISHED(s));
}

/*
 * Start a section.
 */
void
sbuf_start_section(struct sbuf *s, ssize_t *old_lenp)
{

	assert_sbuf_integrity(s);
	assert_sbuf_state(s, 0);

	if (!SBUF_ISSECTION(s)) {
		KASSERT(s->s_sect_len == 0,
		    ("s_sect_len != 0 when starting a section"));
		if (old_lenp != NULL)
			*old_lenp = -1;
		SBUF_SETFLAG(s, SBUF_INSECTION);
	} else {
		KASSERT(old_lenp != NULL,
		    ("s_sect_len should be saved when starting a subsection"));
		*old_lenp = s->s_sect_len;
		s->s_sect_len = 0;
	}
}

/*
 * End the section padding to the specified length with the specified
 * character.
 */
ssize_t
sbuf_end_section(struct sbuf *s, ssize_t old_len, size_t pad, int c)
{
	ssize_t len;

	assert_sbuf_integrity(s);
	assert_sbuf_state(s, 0);
	KASSERT(SBUF_ISSECTION(s),
	    ("attempt to end a section when not in a section"));

	if (pad > 1) {
		len = roundup(s->s_sect_len, pad) - s->s_sect_len;
		for (; s->s_error == 0 && len > 0; len--)
			sbuf_put_byte(s, c);
	}
	len = s->s_sect_len;
	if (old_len == -1) {
		s->s_sect_len = 0;
		SBUF_CLEARFLAG(s, SBUF_INSECTION);
	} else {
		s->s_sect_len += old_len;
	}
	if (s->s_error != 0)
		return (-1);
	return (len);
}
